### PR TITLE
Add Part03 Step03

### LINKED ISSUE #272: Write Part - 3 Step - 3 & unit test
Implement Step03Controller and write all unit tests. Verify the display name of the test conforms to the pattern of "Part X Step Y". Follow the standardize naming, comments, etc. to be consistent.

### PR DESCRIPTION
**Pull request by mmschaefer**
Fix #272 - add part 03 step 03 to the project along with its associated unit testing.  Update part 02 step 11 to be standardized to the project patterns of today.

### CODE REVIEW

File: src/org/etools/j1939_84/controllers/DataRepository.java [hunk 1]
```diff
@@ -12,6 +12,7 @@
 import java.util.Optional;
 import java.util.stream.Collectors;
 import org.etools.j1939_84.bus.j1939.packets.DM11ClearActiveDTCsPacket;
+import org.etools.j1939_84.bus.j1939.packets.DiagnosticTroubleCode;
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/DataRepository.java:15]: This appears to be a stray import.

File: src/org/etools/j1939_84/controllers/part02/Part02Step11Controller.java [hunk 2]
```diff
@@ -52,30 +49,30 @@
...
               TOTAL_STEPS);
-        this.dataRepository = dataRepository;
-        this.diagnosticMessageModule = diagnosticMessageModule;
     }
 
     @Override
     protected void run() throws Throwable {
 
-        diagnosticMessageModule.setJ1939(getJ1939());
+        getDiagnosticMessageModule().setJ1939(getJ1939());
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step11Controller.java]: This line is not necessary when using the `DiagnosticMessageModule` from the super class.

File: src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java [hunk 3]
```diff
@@ -3,17 +3,34 @@
...
+import org.etools.j1939_84.model.OBDModuleInformation;
+import org.etools.j1939_84.model.RequestResult;
 import org.etools.j1939_84.modules.BannerModule;
 import org.etools.j1939_84.modules.DateTimeModule;
 import org.etools.j1939_84.modules.DiagnosticMessageModule;
 import org.etools.j1939_84.modules.EngineSpeedModule;
 import org.etools.j1939_84.modules.VehicleInformationModule;
 
 /**
+ * @author mmschaefer
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java]: This should be your name.

File: src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java [hunk 4]
```diff
@@ -53,13 +70,73 @@
...
+        obdModule.setLastDM27(packet);
+        getDataRepository().putObdModule(obdModule);
+    }
+
+    private void verifyDtcsAreSame(DM27AllPendingDTCsPacket packet) {
+        List<DiagnosticTroubleCode> packetDtcs = packet.getDtcs();
+        OBDModuleInformation obdModule = getDataRepository().getObdModule(packet.getSourceAddress());
+        List<DiagnosticTroubleCode> obdDtcs = obdModule.getEmissionDTCs();
+        // 6.3.3.2.a Fail if (if supported) no ECU reports the same DTC observed in step 6.3.2.1 in a positive DM27 response.
+        if (packetDtcs.size() != obdDtcs.size() || !packetDtcs.equals(obdDtcs)) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java:111]: It's not need to check the size if you're going to use `equals()` as well.

File: src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java [hunk 5]
```diff
@@ -53,13 +70,73 @@
...
+            addFailure(
+                    "6.3.3.2.a - OBD module " + getAddressName(obdModule.getSourceAddress()) +
+                    " saved dtcs are:" + NL +
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java]: Let's not going into the details of the errors.  It makes the report longer and makes our coding more complicated.  The user can compare the two values size by size to determine what the issue is.

File: src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java [hunk 6]
```diff
@@ -53,13 +70,73 @@
...
+                                    .map(DiagnosticTroubleCode::toString)
+                                    .collect(Collectors.joining(NL)));
+        }
+    }
+    private void checkForAdditionalDtc(DM27AllPendingDTCsPacket packet) {
+        List<DiagnosticTroubleCode> packetDtcs = packet.getDtcs();
+        OBDModuleInformation obdModule = getDataRepository().getObdModule(packet.getSourceAddress());
+        List<DiagnosticTroubleCode> obdDtcs = obdModule.getEmissionDTCs();
+        // 6.3.3.3.a. Warn if (if supported) any ECU additional DTCs are provided than the DTC observed in step 6.3.2.1 in a positive DM27 response.
+        if (packetDtcs.size() != obdDtcs.size()) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java]: nit: Technically, we should only warn if there are _more_ DTCs received.

**Review by gumbelmj — CHANGES_REQUESTED** [review]

File: src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java (hunk 4 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java:111]: I read on StackOverflow that this makes the code execute faster.  Though, we aren't checking tons, speed is still a good thing used in moderation.

Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java:111]: That makes sense.

File: src-test/org/etools/j1939_84/controllers/part03/Part03Step03ControllerTest.java [hunk 7]
```diff
@@ -81,17 +79,15 @@
...
-    private DataRepository dataRepository;
-
     private StepController instance;
 
     @Before
     public void setUp() throws Exception {
         DateTimeModule.setInstance(new TestDateTimeModule());
         DateTimeModule dateTimeModule = DateTimeModule.getInstance();
-        dataRepository = DataRepository.newInstance();
+        DataRepository dataRepository = DataRepository.newInstance();
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/controllers/part03/Part03Step03ControllerTest.java]: This is creating a new instance of the `dataRepository`, which isn't shared with the test.  How is this test passing?

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java [hunk 8]
```diff
@@ -54,12 +59,43 @@
     @Override
     protected void run() throws Throwable {
         // 6.3.16.1.a. Turn the engine off.
+        incrementProgress("Part 3, Step 16 Turn Engine Off and keep the ignition key in the off position");
+        ensureKeyOffEngineOff();
+
         // 6.3.16.1.b. Confirm Fault A is still implanted according to the manufacturer’s instruction.
+        incrementProgress("Waiting for implant of Fault A according to the engine manufacturer's instruction");
+        if (!isDevEnv()) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: I'd recommend pushing `!isDevEnv()` up to the super class.

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java [hunk 9]
```diff
@@ -54,12 +59,43 @@
...
         // 6.3.16.1.g. Wait as indicated by the engine manufacturer’s recommendations for Fault A.
+        if (!isDevEnv()) {
+            waitForEngineStart();
+        }
+    }
+
+    private void waitForManufacturerInterval(){
+        String message = "Please wait for the manufacturer's recommended interval with the key in off position" + NL;
+        message += "Press OK to continue the testing" + NL;
+        displayInstructionAndWait(message, "Part 6.3.16.1.c", WARNING);
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: maybe should put the `isDevEnv()` inside the `displayInstructionAndWait()` method?

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 8 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: is part of the changes introduced in step 16

File: src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java (hunk 9 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: Is part of the step 16 changes.

Comment (gumbelmj) [reply src/org/etools/j1939_84/controllers/part03/Part03Step16Controller.java]: Well, I guess it needs to be removed :)

**Review by gumbelmj — APPROVED** [review]

**Pull request merged**